File: CallX2/feature-chat/src/main/java/com/callx/app/chat/performance/SwipeOptimizer.java
package com.callx.app.chat.performance;

import android.view.View;

import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.SimpleItemAnimator;

/**
 * SwipeOptimizer — NO animation version for performance.
 * Spring-back replaced with instant reset.
 */
public final class SwipeOptimizer {

    private SwipeOptimizer() {}

    public static void setTranslationXSafe(View view, float dx) {
        view.setTranslationX(dx);
    }

    /** Instant reset instead of spring animation */
    public static void springBack(View view) {
        view.setTranslationX(0f);
    }

    public static void disableChangeAnimations(RecyclerView rv) {
        RecyclerView.ItemAnimator animator = rv.getItemAnimator();
        if (animator instanceof SimpleItemAnimator) {
            ((SimpleItemAnimator) animator).setSupportsChangeAnimations(false);
        }
    }

    public static void enableHardwareLayer(View v) {
        v.setLayerType(View.LAYER_TYPE_HARDWARE, null);
    }

    public static void clearHardwareLayer(View v) {
        v.setLayerType(View.LAYER_TYPE_NONE, null);
    }
}

File: CallX2/feature-chat/src/main/java/com/callx/app/chat/performance/SwipeOptimizer.kt
package com.callx.app.chat.performance

import android.view.View
import androidx.dynamicanimation.animation.DynamicAnimation
import androidx.dynamicanimation.animation.SpringAnimation
import androidx.dynamicanimation.animation.SpringForce
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.SimpleItemAnimator

/**
 * SwipeOptimizer — Performance utilities for lag-free swipe-to-reply.
 *
 * Key principles:
 *   • ONLY translationX during swipe — never requestLayout()
 *   • Disable RecyclerView change animations (prevents flicker)
 *   • Spring-back uses DynamicAnimation (hardware-accelerated)
 */
object SwipeOptimizer {

    fun setTranslationXSafe(view: View, dx: Float) {
        view.translationX = dx
    }

    fun springBack(view: View) {
        val currentTx = view.translationX
        if (Math.abs(currentTx) < 0.5f) {
            view.translationX = 0f
            return
        }
        SpringAnimation(view, DynamicAnimation.TRANSLATION_X, 0f).apply {
            spring.stiffness = SpringForce.STIFFNESS_MEDIUM
            spring.dampingRatio = 0.7f
            setStartVelocity(-currentTx * 2f)
            start()
        }
    }

    fun disableChangeAnimations(rv: RecyclerView) {
        (rv.itemAnimator as? SimpleItemAnimator)?.supportsChangeAnimations = false
    }

    fun enableHardwareLayer(v: View) {
        v.setLayerType(View.LAYER_TYPE_HARDWARE, null)
    }

    fun clearHardwareLayer(v: View) {
        v.setLayerType(View.LAYER_TYPE_NONE, null)
    }
}
